### DISCUSSION TITLE
Gossip Layer

### DISCUSSION THREAD

**Discussion by cmwaters** [tendermint/tendermint#6722] [General]
Following on from @milosevic's call and largely spurred by conversations with @tychoish, I've been thinking about the dynamic between the reactors and the p2p layer and I would like to open a discussion with the rest of the team on this relationship. First off, to me, _Gossip_ is the method with which information is disseminated throughout a network. With this definition, a naive method would be to connect with all peers and broadcast each message to all of them. This has a network load of n<sup>2</sup>. 

Thinking about the naive method highlights some desirable properties of a good gossip method:
-  We want to send the minimal amount of messages possible
-  We want fault tolerance and protection from eclipse attacks

Sam and I were discussing some minor issue (https://github.com/tendermint/tendermint/issues/6598) whereby reactors may receive peer updates but that peer might not be receiving on that channel i.e. the statesync reactor adds a peer to ask for snapshots but that peer isn't running state sync. It's minor because the router will intervene and stop it but it still means that the reactor has a slightly invalid model of the network. Sam then asked the question: why does the reactor need to know about peers? This question strikes me as being very core to the ideas Zarko proposed: reactors should only entail the logic of the protocol (i.e. for state sync this is passing the application snapshot from one node to another) and perhaps they don't need to know the method with which the information gets disseminated around the network. The gossiping pattern across different reactors are very similar: We want an efficient means of passing information from producers to consumers. In the mempool, a producer is someone who receives a TX via the RPC and the consumers are the list of proposers (or all validators). In consensus, the producer can be a validator who signed a vote for a block and the consumers are all the other validators who need the vote to help them reach consensus. 

The way most reactors achieve this implementation is by sending extra messages which lets peers know what information they have and what information they are missing. Most reactors therefore keep their own peer state. If this is a generic pattern, then would it make sense to have this as part of the p2p layer. What are people's thoughts?

**Comment by alexanderbez** [comment]
> We want fault tolerance and protection from eclipse attacks

@cmwaters please either explain or provide reference(s) to eclipse attacks for future readers who may not be aware what they are.

I'm also having a bit of trouble grasping the concrete proposal or question. Can you summarize in a sentence or two or maybe just point it out?

> The way most reactors achieve this implementation is by sending extra messages which lets peers know what information they have and what information they are missing.

Note, this model is essentially how Ethereum's mempool gossiping is designed and works pretty efficiently -- gossip a light data structure of what you have, the receiving peer asks for the diff in return.

**Comment by tychoish** [comment]
I think the general idea of "reactors shouldn't need to depend on the peer manager" is a good goal, and I think there are ways that we can improve on this model. 

I think most communications between reactors don't actually need to know things about specific peers, as messages are:
- broadcast, send to everyone,
- reply to a specific incoming request
- send to all peers that match a specific characteristic/tag/etc. 

I think the message sync ("gossip layer") can mostly handle those routing details and we can (iteratively,) remove them from the reactors. There's a number of places where reactors do some things that impact peer scoring, and can mark a peer bad, which are a bit harder to account for, but I think it's possible.

I think I might summarize the current state of the reactors as "thinking a lot about where the messages it's sending are going" and I think adding a more bulletproof abstraction for reactors between the p2p layer and the reactor logic could be really good for the overall code health and organization. And I think it would make it easier to extend, test, and maintain the system. 

There are a lot of details that are very abstract here, both in terms of the implementation and how we might iterate there, but I think the big take away of thinking about "gossip" as a separate conceptual/functional layer from p2p (peer management, communication) from the reactors (business logic,) is a good high level model.

**Comment by williambanfield** [comment]
I wonder if maybe there are two types of communication being discussed here that might need to be handled separately? 

1. Gossiped messages
2. point-to-point messages

The consensus protocol messages would fall under (1), whereas other messages may fall under (2). Gossiped messages need to deliver to all validators, whether or not there is a direct point-to-point line between the peer sending and the peer receiving the message. Point-to-point messages are more typical request-response messages. An example of this is during sync'ing where the dispatcher asks a specific peer for a specific lightblock. 

Both of these types of messages rely on having a pool of 'peers' that they can connect to. Gossiping adds the additional piece of a validator saying "I just received message X, do all of my peers have message X"?

**Reply by alexanderbez** [reply]
Yeah that makes sense to make that distinction. In the current context of our architecture, the difference in what you're describing is sending envelopes with `broadcast: true`, where `true` means gossip this envelope to all my known connected peers.
